File: src/main/java/com/dude/dms/ui/builder/tags/TagBuilderFactory.java
package com.dude.dms.ui.builder.tags;

import com.dude.dms.backend.data.Tag;
import com.dude.dms.backend.service.DocService;
import com.dude.dms.backend.service.TagService;
import com.dude.dms.ui.builder.BuilderFactory;
import com.dude.dms.ui.builder.Factory;

public class TagBuilderFactory extends Factory {

    private final TagService tagService;
    private final DocService docService;

    public TagBuilderFactory(BuilderFactory builderFactory, TagService tagService, DocService docService) {
        super(builderFactory);
        this.tagService = tagService;
        this.docService = docService;
    }

    public TagSearchResultBuilder searchResult(Tag tag) {
        return new TagSearchResultBuilder(tag, builderFactory.tags().editDialog(tag).build(), docService);
    }

    public TagCreateDialogBuilder createDialog() {
        return new TagCreateDialogBuilder(builderFactory, tagService);
    }

    public TagEditDialogBuilder editDialog(Tag tag) {
        return new TagEditDialogBuilder(builderFactory, tag, tagService);
    }

    public TagSelectorBuilder selector() {
        return new TagSelectorBuilder(tagService);
    }
}

File: src/main/java/com/dude/dms/ui/builder/tags/TagBuilderFactory.kt
package com.dude.dms.ui.builder.tags

import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.service.*
import com.dude.dms.ui.builder.BuilderFactory
import com.dude.dms.ui.builder.Factory

class TagBuilderFactory(
        builderFactory: BuilderFactory,
        private val tagService: TagService,
        private val docService: DocService,
        private val mailService: MailService,
        private val attributeService: AttributeService,
        private val plainTextRuleService: PlainTextRuleService,
        private val regexRuleService: RegexRuleService,
        private val mailFilterService: MailFilterService
) : Factory(builderFactory) {

    fun searchResult(tag: Tag) = TagSearchResultBuilder(tag, builderFactory.tags().editDialog(tag).build(), docService)

    fun createDialog() = TagCreateDialogBuilder(builderFactory, tagService)

    fun editDialog(tag: Tag) = TagEditDialogBuilder(builderFactory, tag, tagService, docService)

    fun deleteDialog(tag: Tag) = TagDeleteDialogBuilder(tag, tagService, docService, mailService, attributeService, plainTextRuleService, regexRuleService, mailFilterService)

    fun selector() = TagSelectorBuilder(tagService)
}